File: lpp/lexer.py
from lpp.token import Token, TokenType
from re import match

class Lexer:
    def __init__(self, source: str) -> None:
        self._source: str = source
        self._character: str = ''
        self._read_position: int = 0
        self._position: int = 0
        self._read_character()

    def next_token(self) -> Token:
        if match(r'^=$', self._character):
            token = Token(TokenType.ASSING, self._character)
        elif match(r'^\+$', self._character):
            token = Token(TokenType.PLUS, self._character)
        elif match(r'^$', self._character):
            token = Token(TokenType.EOF, self._character)
        elif match(r'^\($', self._character):
            token = Token(TokenType.LPAREN, self._character)
        elif match(r'^\)$', self._character):
            token = Token(TokenType.RPAREN, self._character)
        elif match(r'^\{$', self._character):
            token = Token(TokenType.LBRACE, self._character)
        elif match(r'^\}$', self._character):
            token = Token(TokenType.RBRACE, self._character)
        elif match(r'^\,$', self._character):
            token = Token(TokenType.COMMA, self._character)
        elif match(r'^\;$', self._character):
            token = Token(TokenType.SEMICOLON, self._character)
        else:
            token = Token(TokenType.ILLEGAL, self._character)

        self._read_character()
        return token

    def _read_character(self) -> None:
        if self._read_position >= len(self._source):
            self._character = ''
        else:
            self._character = self._source[self._read_position]

        self._position = self._read_position
        self._read_position += 1

File: lpp/lexer.go
package lpp

import (
	"regexp"
	"unicode/utf8"
)

type Lexer struct {
	source        string
	character     string
	read_position int
	position      int
}

// create a new lexer
func NewLexer(source string) *Lexer {
	lexer := &Lexer{
		source:        source,
		character:     "",
		read_position: 0,
		position:      0,
	}

	lexer.readCharacter()
	return lexer
}

// read next token
func (l *Lexer) NextToken() Token {
	l.skipWhiteSpaces()
	var token Token

	if equal, _ := regexp.MatchString(`^=$`, l.character); equal {
		token = Token{Token_type: ASSING, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^\+$`, l.character); equal {
		token = Token{Token_type: PLUS, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^$`, l.character); equal {
		token = Token{Token_type: EOF, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^\($`, l.character); equal {
		token = Token{Token_type: LPAREN, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^\)$`, l.character); equal {
		token = Token{Token_type: RPAREN, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^\{$`, l.character); equal {
		token = Token{Token_type: LBRACE, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^\}$`, l.character); equal {
		token = Token{Token_type: RBRACE, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^,$`, l.character); equal {
		token = Token{Token_type: COMMA, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^;$`, l.character); equal {
		token = Token{Token_type: SEMICOLON, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^<$`, l.character); equal {
		token = Token{Token_type: LT, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^>$`, l.character); equal {
		token = Token{Token_type: GT, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^\-$`, l.character); equal {
		token = Token{Token_type: MINUS, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^\/$`, l.character); equal {
		token = Token{Token_type: DIVISION, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^\*$`, l.character); equal {
		token = Token{Token_type: TIMES, Literal: l.character}

	} else if equal, _ := regexp.MatchString(`^\!$`, l.character); equal {
		token = Token{Token_type: NOT, Literal: l.character}

	} else if l.isLetter(l.character) {
		literal := l.readIdentifier()
		token_type := LookUpTokenType(literal)
		return Token{Token_type: token_type, Literal: literal}

	} else if l.isNumber(l.character) {
		literal := l.readNumber()
		return Token{Token_type: INT, Literal: literal}

	} else {
		token = Token{Token_type: ILLEGAL, Literal: l.character}
	}

	l.readCharacter()
	return token
}

// check if current character is letter
func (l *Lexer) isLetter(char string) bool {
	isValid, _ := regexp.MatchString(`^[a-záéíóúA-ZÁÉÍÓÚñÑ_]$`, char)
	return isValid
}

// check if current character is number
func (l *Lexer) isNumber(char string) bool {
	isValid, _ := regexp.MatchString(`^\d$`, char)
	return isValid
}

// read current character.
func (l *Lexer) readCharacter() {
	if l.read_position >= utf8.RuneCountInString(l.source) {
		l.character = ""
	} else {
		l.character = string([]rune(l.source)[l.read_position])
	}

	l.position = l.read_position
	l.read_position++
}

// read character sequence
func (l *Lexer) readIdentifier() string {
	initialPosition := l.position
	for l.isLetter(l.character) {
		l.readCharacter()
	}

	return string(l.source[initialPosition:l.position])
}

// read number sequence of characters
func (l *Lexer) readNumber() string {
	initialPosition := l.position
	for l.isNumber(l.character) {
		l.readCharacter()
	}

	return string(l.source[initialPosition:l.position])
}

// skipp whitespaces
func (l *Lexer) skipWhiteSpaces() {
	m, _ := regexp.Compile(`^\s$`)
	for m.MatchString(l.character) {
		l.readCharacter()
	}
}
